File: examples/crypto/sha256/merkle-tree.ts
/**
 * Merkle Tree with SHA256 Example
 *
 * Demonstrates building authenticated data structures:
 * - Merkle tree construction
 * - Merkle proofs
 * - Verification
 * - Bitcoin-style double hashing
 */

import { SHA256 } from '../../../src/crypto/sha256/SHA256.js';

console.log('=== SHA256 Merkle Tree ===\n');

// Helper: Double SHA-256 (Bitcoin style)
function doubleSha256(data: Uint8Array): Uint8Array {
  return SHA256.hash(SHA256.hash(data));
}

// Build Merkle tree from leaves
function buildMerkleTree(leaves: Uint8Array[]): Uint8Array[][] {
  if (leaves.length === 0) {
    throw new Error('Cannot build tree from empty leaves');
  }

  const tree: Uint8Array[][] = [];

  // First level: hash all leaves
  tree.push(leaves.map(leaf => doubleSha256(leaf)));

  // Build subsequent levels
  let currentLevel = tree[0];
  while (currentLevel.length > 1) {
    const nextLevel: Uint8Array[] = [];

    for (let i = 0; i < currentLevel.length; i += 2) {
      const left = currentLevel[i];
      const right = currentLevel[i + 1] || left; // Duplicate if odd number

      // Combine and hash
      const combined = new Uint8Array(64);
      combined.set(left, 0);
      combined.set(right, 32);
      nextLevel.push(doubleSha256(combined));
    }

    tree.push(nextLevel);
    currentLevel = nextLevel;
  }

  return tree;
}

// Get Merkle root (top of tree)
function getMerkleRoot(tree: Uint8Array[][]): Uint8Array {
  return tree[tree.length - 1][0];
}

// Generate Merkle proof for a leaf
function generateMerkleProof(tree: Uint8Array[][], leafIndex: number): { hash: Uint8Array; position: 'left' | 'right' }[] {
  const proof: { hash: Uint8Array; position: 'left' | 'right' }[] = [];
  let index = leafIndex;

  for (let level = 0; level < tree.length - 1; level++) {
    const isRightNode = index % 2 === 1;
    const siblingIndex = isRightNode ? index - 1 : index + 1;

    if (siblingIndex < tree[level].length) {
      proof.push({
        hash: tree[level][siblingIndex],
        position: isRightNode ? 'left' : 'right',
      });
    }

    index = Math.floor(index / 2);
  }

  return proof;
}

// Verify Merkle proof
function verifyMerkleProof(
  leaf: Uint8Array,
  proof: { hash: Uint8Array; position: 'left' | 'right' }[],
  root: Uint8Array
): boolean {
  let hash = doubleSha256(leaf);

  for (const { hash: siblingHash, position } of proof) {
    const combined = new Uint8Array(64);
    if (position === 'left') {
      combined.set(siblingHash, 0);
      combined.set(hash, 32);
    } else {
      combined.set(hash, 0);
      combined.set(siblingHash, 32);
    }
    hash = doubleSha256(combined);
  }

  return SHA256.toHex(hash) === SHA256.toHex(root);
}

// Example 1: Build simple Merkle tree
console.log('1. Building Merkle Tree');
console.log('-'.repeat(70));

const leaves = [
  new TextEncoder().encode('Transaction 1'),
  new TextEncoder().encode('Transaction 2'),
  new TextEncoder().encode('Transaction 3'),
  new TextEncoder().encode('Transaction 4'),
];

console.log('Building tree from', leaves.length, 'leaves:\n');
leaves.forEach((leaf, i) => {
  console.log(`  Leaf ${i}: "${new TextDecoder().decode(leaf)}"`);
});
console.log();

const tree = buildMerkleTree(leaves);
const root = getMerkleRoot(tree);

console.log('Tree structure:');
for (let level = tree.length - 1; level >= 0; level--) {
  const indent = ' '.repeat(level * 2);
  console.log(`Level ${tree.length - 1 - level}:${indent}`, tree[level].length, 'nodes');
}
console.log();
console.log('Merkle Root:', SHA256.toHex(root));
console.log();

// Example 2: Merkle proof generation and verification
console.log('2. Merkle Proof Generation and Verification');
console.log('-'.repeat(70));

const leafIndex = 2; // Prove "Transaction 3"
const leaf = leaves[leafIndex];

console.log(`Generating proof for leaf ${leafIndex}: "${new TextDecoder().decode(leaf)}"\n`);

const proof = generateMerkleProof(tree, leafIndex);

console.log('Proof path:');
proof.forEach((step, i) => {
  console.log(`  Step ${i + 1}: ${step.position.padEnd(5)} 0x${SHA256.toHex(step.hash).slice(2, 18)}...`);
});
console.log();

const isValid = verifyMerkleProof(leaf, proof, root);
console.log('Proof verification:', isValid ? '✓ VALID' : '✗ INVALID');
console.log();

// Tamper with leaf
const tamperedLeaf = new TextEncoder().encode('Transaction 3 (modified)');
const tamperedValid = verifyMerkleProof(tamperedLeaf, proof, root);
console.log('Tampered leaf verification:', tamperedValid ? '✓ VALID' : '✗ INVALID');
console.log();

// Example 3: Odd number of leaves
console.log('3. Merkle Tree with Odd Number of Leaves');
console.log('-'.repeat(70));

const oddLeaves = [
  new TextEncoder().encode('TX 1'),
  new TextEncoder().encode('TX 2'),
  new TextEncoder().encode('TX 3'),
  new TextEncoder().encode('TX 4'),
  new TextEncoder().encode('TX 5'),
];

console.log('Building tree from', oddLeaves.length, 'leaves (odd number)');
console.log('Last node will be duplicated at each level\n');

const oddTree = buildMerkleTree(oddLeaves);
const oddRoot = getMerkleRoot(oddTree);

for (let level = 0; level < oddTree.length; level++) {
  console.log(`Level ${level}:`, oddTree[level].length, 'nodes');
}
console.log();
console.log('Merkle Root:', SHA256.toHex(oddRoot));
console.log();

// Example 4: Large tree (simulated Bitcoin block)
console.log('4. Large Merkle Tree (Simulated Bitcoin Block)');
console.log('-'.repeat(70));

const txCount = 1000;
const largeTxs: Uint8Array[] = [];

for (let i = 0; i < txCount; i++) {
  const tx = new TextEncoder().encode(`Transaction ${i}`);
  largeTxs.push(tx);
}

console.log(`Building Merkle tree for ${txCount} transactions...\n`);

const start = performance.now();
const largeTree = buildMerkleTree(largeTxs);
const largeRoot = getMerkleRoot(largeTree);
const elapsed = performance.now() - start;

console.log('Tree levels:', largeTree.length);
console.log('Merkle root:', SHA256.toHex(largeRoot));
console.log('Build time:', elapsed.toFixed(2), 'ms');
console.log();

// Proof size scales logarithmically
const proofIndex = 500;
const largeProof = generateMerkleProof(largeTree, proofIndex);
console.log('Proof size for transaction', proofIndex, ':', largeProof.length, 'hashes');
console.log('Proof size scales O(log n) - for 1000 txs, only', largeProof.length, 'hashes needed!');
console.log();

// Example 5: Merkle tree efficiency comparison
console.log('5. Efficiency Analysis');
console.log('-'.repeat(70));

const sizes = [10, 100, 1000, 10000];

console.log('Transaction Count | Tree Levels | Proof Size | Space Savings');
console.log('-'.repeat(70));

for (const size of sizes) {
  const txs = Array.from({ length: size }, (_, i) =>
    new TextEncoder().encode(`TX ${i}`)
  );
  const t = buildMerkleTree(txs);
  const p = generateMerkleProof(t, 0);

  const withoutMerkle = size * 32; // All hashes
  const withMerkle = p.length * 32; // Only proof hashes
  const savings = ((1 - withMerkle / withoutMerkle) * 100).toFixed(1);

  console.log(
    `${size.toString().padStart(17)} | ${t.length.toString().padStart(11)} | ${p.length.toString().padStart(10)} | ${savings.padStart(13)}%`
  );
}
console.log();

// Example 6: Merkle tree visualization (small tree)
console.log('6. Merkle Tree Visualization');
console.log('-'.repeat(70));

const smallLeaves = [
  new TextEncoder().encode('A'),
  new TextEncoder().encode('B'),
  new TextEncoder().encode('C'),
  new TextEncoder().encode('D'),
];

const smallTree = buildMerkleTree(smallLeaves);

console.log('Tree structure:\n');
console.log('                Root');
console.log('                 |');
console.log('         +-------+-------+');
console.log('         |               |');
console.log('       Hash1           Hash2');
console.log('       /   \\           /   \\');
console.log('      /     \\         /     \\');
console.log('     A       B       C       D');
console.log();

for (let level = smallTree.length - 1; level >= 0; level--) {
  console.log(`Level ${smallTree.length - 1 - level}:`);
  smallTree[level].forEach((hash, i) => {
    console.log(`  Node ${i}: 0x${SHA256.toHex(hash).slice(2, 18)}...`);
  });
  console.log();
}

console.log('=== Merkle Tree Complete ===');

File: examples/crypto/sha256/merkle-tree.zig
const std = @import("std");
const crypto = @import("crypto");
const SHA256 = crypto.SHA256;

pub fn main() !void {
    var gpa = std.heap.GeneralPurposeAllocator(.{}){};
    defer _ = gpa.deinit();
    const allocator = gpa.allocator();

    std.debug.print("=== SHA256 Merkle Tree ===\n\n", .{});

    // Example 1: Build simple Merkle tree
    std.debug.print("1. Building Merkle Tree\n", .{});
    std.debug.print("{s}\n", .{"-" ** 70});

    const leaves = [_][]const u8{
        "Transaction 1",
        "Transaction 2",
        "Transaction 3",
        "Transaction 4",
    };

    std.debug.print("Building tree from {} leaves:\n\n", .{leaves.len});
    for (leaves, 0..) |leaf, i| {
        std.debug.print("  Leaf {}: \"{s}\"\n", .{ i, leaf });
    }
    std.debug.print("\n", .{});

    var tree = try buildMerkleTree(&leaves, allocator);
    defer {
        for (tree.items) |level| {
            allocator.free(level);
        }
        tree.deinit(allocator);
    }

    const root = tree.items[tree.items.len - 1][0];

    std.debug.print("Tree structure:\n", .{});
    var level: usize = tree.items.len;
    while (level > 0) : (level -= 1) {
        const indent = " " ** ((level - 1) * 2);
        std.debug.print("Level {}:{s} {} nodes\n", .{
            tree.items.len - level,
            indent,
            tree.items[level - 1].len / 32,
        });
    }
    std.debug.print("\n", .{});
    std.debug.print("Merkle Root: 0x{s}\n\n", .{std.fmt.fmtSliceHexLower(root[0..32])});

    // Example 2: Merkle proof generation and verification
    std.debug.print("2. Merkle Proof Generation and Verification\n", .{});
    std.debug.print("{s}\n", .{"-" ** 70});

    const leaf_index: usize = 2;
    const leaf = leaves[leaf_index];

    std.debug.print("Generating proof for leaf {}: \"{s}\"\n\n", .{ leaf_index, leaf });

    const proof = try generateMerkleProof(tree, leaf_index, allocator);
    defer allocator.free(proof);

    std.debug.print("Proof path:\n", .{});
    for (proof, 0..) |step, i| {
        const pos_str = if (step.position == .left) "left " else "right";
        std.debug.print("  Step {}: {s} 0x{s}...\n", .{
            i + 1,
            pos_str,
            std.fmt.fmtSliceHexLower(step.hash[0..8]),
        });
    }
    std.debug.print("\n", .{});

    const is_valid = try verifyMerkleProof(leaf, proof, root[0..32], allocator);
    const valid_str = if (is_valid) "✓ VALID" else "✗ INVALID";
    std.debug.print("Proof verification: {s}\n\n", .{valid_str});

    // Tamper with leaf
    const tampered_leaf = "Transaction 3 (modified)";
    const tampered_valid = try verifyMerkleProof(tampered_leaf, proof, root[0..32], allocator);
    const tampered_str = if (tampered_valid) "✓ VALID" else "✗ INVALID";
    std.debug.print("Tampered leaf verification: {s}\n\n", .{tampered_str});

    // Example 3: Odd number of leaves
    std.debug.print("3. Merkle Tree with Odd Number of Leaves\n", .{});
    std.debug.print("{s}\n", .{"-" ** 70});

    const odd_leaves = [_][]const u8{
        "TX 1",
        "TX 2",
        "TX 3",
        "TX 4",
        "TX 5",
    };

    std.debug.print("Building tree from {} leaves (odd number)\n", .{odd_leaves.len});
    std.debug.print("Last node will be duplicated at each level\n\n", .{});

    var odd_tree = try buildMerkleTree(&odd_leaves, allocator);
    defer {
        for (odd_tree.items) |lev| {
            allocator.free(lev);
        }
        odd_tree.deinit(allocator);
    }

    const odd_root = odd_tree.items[odd_tree.items.len - 1][0];

    for (odd_tree.items, 0..) |lev, i| {
        std.debug.print("Level {}: {} nodes\n", .{ i, lev.len / 32 });
    }
    std.debug.print("\n", .{});
    std.debug.print("Merkle Root: 0x{s}\n\n", .{std.fmt.fmtSliceHexLower(odd_root[0..32])});

    std.debug.print("=== Merkle Tree Complete ===\n", .{});
}

// Helper: Double SHA-256 (Bitcoin style)
fn doubleSha256(data: []const u8) [32]u8 {
    const first = SHA256.hash(data);
    return SHA256.hash(&first);
}

const ProofStep = struct {
    hash: [32]u8,
    position: enum { left, right },
};

// Build Merkle tree from leaves
fn buildMerkleTree(leaves: []const []const u8, allocator: std.mem.Allocator) !std.ArrayList([]u8) {
    var tree = std.ArrayList([]u8){};

    // First level: hash all leaves
    var first_level = try allocator.alloc(u8, leaves.len * 32);
    for (leaves, 0..) |leaf, i| {
        const hash = doubleSha256(leaf);
        @memcpy(first_level[i * 32 .. (i + 1) * 32], &hash);
    }
    try tree.append(allocator, first_level);

    // Build subsequent levels
    var current_level = first_level;
    while (current_level.len > 32) {
        const node_count = current_level.len / 32;
        const next_node_count = (node_count + 1) / 2;
        var next_level = try allocator.alloc(u8, next_node_count * 32);

        var i: usize = 0;
        while (i < node_count) : (i += 2) {
            const left = current_level[i * 32 .. (i + 1) * 32];
            const right = if (i + 1 < node_count)
                current_level[(i + 1) * 32 .. (i + 2) * 32]
            else
                left;

            // Combine and hash
            var combined: [64]u8 = undefined;
            @memcpy(combined[0..32], left);
            @memcpy(combined[32..64], right);
            const hash = doubleSha256(&combined);
            @memcpy(next_level[(i / 2) * 32 .. ((i / 2) + 1) * 32], &hash);
        }

        try tree.append(allocator, next_level);
        current_level = next_level;
    }

    return tree;
}

// Generate Merkle proof for a leaf
fn generateMerkleProof(tree: std.ArrayList([]u8), leaf_index: usize, allocator: std.mem.Allocator) ![]ProofStep {
    var proof = std.ArrayList(ProofStep){};
    defer proof.deinit(allocator);

    var index = leaf_index;

    for (tree.items[0 .. tree.items.len - 1]) |level| {
        const node_count = level.len / 32;
        const is_right_node = index % 2 == 1;
        const sibling_index = if (is_right_node) index - 1 else index + 1;

        if (sibling_index < node_count) {
            var step: ProofStep = undefined;
            @memcpy(&step.hash, level[sibling_index * 32 .. (sibling_index + 1) * 32]);
            step.position = if (is_right_node) .left else .right;
            try proof.append(allocator, step);
        }

        index = index / 2;
    }

    return proof.toOwnedSlice(allocator);
}

// Verify Merkle proof
fn verifyMerkleProof(
    leaf: []const u8,
    proof: []const ProofStep,
    root: []const u8,
    allocator: std.mem.Allocator,
) !bool {
    _ = allocator;
    var hash = doubleSha256(leaf);

    for (proof) |step| {
        var combined: [64]u8 = undefined;
        if (step.position == .left) {
            @memcpy(combined[0..32], &step.hash);
            @memcpy(combined[32..64], &hash);
        } else {
            @memcpy(combined[0..32], &hash);
            @memcpy(combined[32..64], &step.hash);
        }
        hash = doubleSha256(&combined);
    }

    return std.mem.eql(u8, &hash, root);
}
